utility/tests/test_repoutil.py

```diff
@@ -0,0 +1,260 @@
+		self.assertEqual(self.repo1.updateSubmodule(False, False), None)  
+		
+		# the repo repo5 is with submodules
+		self.assertEqual(self.repo5.clone(self.myname5, self.myemail5), None)  
+		repodir25 = self.repo5.localdir + '/'
+		self.assertTrue(os.path.isfile(repodir25 + '.gitmodules'))
+
+		config = configparser.ConfigParser()
+		config.read(repodir25 + '.gitmodules')
+		# use rmdir to ascertain that submodules' folders are present and empty
+		for module in config.sections():
+			self.assertTrue(os.path.exists(repodir25 + config[module]['path']), repodir25)
+			os.rmdir(repodir25 + config[module]['path'])
+		# recreate previously deleted submodules' folders
+		for module in config.sections():
+			self.assertFalse(os.path.exists(repodir25 + config[module]['path']), repodir25)
+			os.mkdir(repodir25 + config[module]['path'])
+			self.assertTrue(os.path.exists(repodir25 + config[module]['path']), repodir25)
+
+		# Init = False; should not update submodules
+		self.assertEqual(self.repo5.updateSubmodule(True, False), None)  
+		self.assertEqual(self.repo5.updateSubmodule(False, False), None)  
+
+		# use rmdir to ascertain that submodules' folders are present and empty
+		for module in config.sections():
+			self.assertTrue(os.path.exists(repodir25 + config[module]['path']), repodir25)
+			os.rmdir(repodir25 + config[module]['path'])
+		# recreate previously deleted submodules' folders
+		for module in config.sections():
+			self.assertFalse(os.path.exists(repodir25 + config[module]['path']), repodir25)
+			os.mkdir(repodir25 + config[module]['path'])
+			self.assertTrue(os.path.exists(repodir25 + config[module]['path']), repodir25)
+
+		# the key test - how updateSubmodule works ...
+		self.assertEqual(self.repo5.updateSubmodule(), None)  
+
+		# use rmdir to ascetain that after update the submodules' folders are present and contains git repos - it is expected that rmdir would fail if it is applied on non empty folder
+		for module in config.sections():
+			self.assertTrue(os.path.exists(repodir25 + config[module]['path']), repodir25)
+			with self.assertRaises(OSError):
+				os.rmdir(repodir25 + config[module]['path'])
+
+			self.assertTrue(repodir25 + config[module]['path'] + '/.git')
+
+		self.assertEqual(self.repo5.updateSubmodule(True, True), None)  
+		self.assertEqual(self.repo5.updateSubmodule(False, True), None)  
+			
+	def test_get_commit_hash(self):
+		# the repo repo5 is with submodules
+		self.assertEqual(self.repo5.clone(self.myname5, self.myemail5), None)  
+		repodir25 = self.repo5.localdir + '/'
+		self.assertTrue(os.path.isfile(repodir25 + '.gitmodules'))
+		
+		self.assertEqual(self.repo5.get_commit_hash(), '007fdc8e7d9c6ff70f2c9624c68aa83ef993b45a')
+		self.assertEqual(self.repo5.get_commit_hash(None, 'master'), '007fdc8e7d9c6ff70f2c9624c68aa83ef993b45a')
+		self.assertEqual(self.repo5.get_commit_hash('backend','master'), 'd1196aa44777fea39890a453d676a8775f4b53dc')
```

Comment:
I found problem with my code in here.. it is updated and should work right now. But there is a problem. since you are using submodule backend and master branch (which is always updated) it will not work like this. I created a branch tests which will not be deleted or changed. please use that commit id and branch.

Reply:
Need to cast variable of type Head to string
            for ref in refs:
                if str(ref) == branch or str(ref) == 'origin/{}'.format(branch):
Then it works

Reply:
you are right. thank you
I used name variable instead.
fixed